magenta/js/src/sequences.ts

```diff
@@ -0,0 +1,272 @@
+   *    quantized time steps.
+   */
+  private static quantizeNotes(ns:NoteSequence, stepsPerSecond:number) {
+    for(const note of ns.notes) {
+      // Quantize the start and end times of the note.
+      note.quantizedStartStep = this.quantizeToStep(
+          note.startTime, stepsPerSecond);
+      note.quantizedEndStep = this.quantizeToStep(
+          note.endTime, stepsPerSecond);
+      if(note.quantizedEndStep === note.quantizedStartStep) {
+          note.quantizedEndStep = note.quantizedEndStep + 1;
+      }
+
+      // Do not allow notes to start or end in negative time.
+      if(note.quantizedStartStep < 0 || note.quantizedEndStep < 0) {
+          throw new NegativeTimeException(
+              `Got negative note time: start_step = ` +
+              `${note.quantizedStartStep}, end_step = ` +
+              `${note.quantizedEndStep}`);
+      }
+
+      // Extend quantized sequence if necessary.
+      if(note.quantizedEndStep > ns.totalQuantizedSteps) {
+          ns.totalQuantizedSteps = note.quantizedEndStep;
+      }
+    }
+
+    // Also quantize control changes and text annotations.
+    for(const event of ns.controlChanges.concat(ns.textAnnotations)) {
+      // Quantize the event time, disallowing negative time.
+      event.quantizedStep = this.quantizeToStep(event.time, stepsPerSecond);
+      if(event.quantizedStep < 0) {
+        throw new NegativeTimeException(
+            `Got negative event time: step = ${event.quantizedStep}`);
+      }
+    }
+  }
+
+  /**
+   * Quantize a NoteSequence proto relative to tempo.
+   *
+   * The input NoteSequence is copied and quantization-related fields are
+   * populated. Sets the `steps_per_quarter` field in the `quantization_info`
+   * message in the NoteSequence.
+   *
+   * Note start and end times, and chord times are snapped to a nearby quantized
+   * step, and the resulting times are stored in a separate field (e.g.,
+   * QuantizedStartStep). See the comments above `QUANTIZE_CUTOFF` for details
+   * on how the quantizing algorithm works.
+   *
+   * @param note_sequence A music_pb2.NoteSequence protocol buffer.
+   * @param steps_per_quarter Each quarter note of music will be divided into
+   *    this many quantized time steps.
+   * @returns A copy of the original NoteSequence, with quantized times added.
+   *
+   * @throws {MultipleTempoException} If there is a change in tempo in
+   *    note sequence.
+   * @throws {MultipleTimeSignatureException} If there is a change in time
+   *    signature in noteSequence.
+   * @throws {BadTimeSignatureException} If the time signature found in
+   *    noteSequence has a 0 numerator or a denominator which is not a power of
+   *    2.
+   * @throws {NegativeTimeException} If a note or chord occurs at a negative
+   *    time.
+   */
+  public static quantizeNoteSequence(noteSequence: NoteSequence,
+      stepsPerQuarter: number): NoteSequence {
+    // Make a copy.
+    const qns = NoteSequence.decode(
```

Comment:
Is this really the best way to make a copy??

Reply:
If so, can we put this in a helper fn?

Reply:
@hapticdata do you have thoughts on a better way to do this? I got this idea from this thread: https://github.com/dcodeIO/protobuf.js/issues/315